In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("revivepay_historical_payments.csv")

In [3]:
df.head()

,payment_id,customer_id,amount_inr,payment_method,failure_type,hour_of_day,minutes_since_failure,retry_count,previous_payments,previous_successes,previous_success_rate,recovered
0,PAY_00001,CUST_2329,417.77,upi,checkout_abandoned,13,1031,2,5,2,0.40,0
1,PAY_00002,CUST_1521,1312.63,upi,insufficient_funds,16,1527,2,4,1,0.25,0
2,PAY_00003,CUST_1394,444.44,card,checkout_abandoned,17,1557,1,2,1,0.50,0
3,PAY_00004,CUST_1558,9216.71,card,checkout_abandoned,9,1021,2,4,2,0.50,0
4,PAY_00005,CUST_1934,2133.53,card,card_declined,4,793,0,2,1,0.50,0


In [5]:
df.shape

(10000, 12)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   payment_id             10000 non-null  object 
 1   customer_id            10000 non-null  object 
 2   amount_inr             10000 non-null  float64
 3   payment_method         10000 non-null  object 
 4   failure_type           10000 non-null  object 
 5   hour_of_day            10000 non-null  int64  
 6   minutes_since_failure  10000 non-null  int64  
 7   retry_count            10000 non-null  int64  
 8   previous_payments      10000 non-null  int64  
 9   previous_successes     10000 non-null  int64  
 10  previous_success_rate  10000 non-null  float64
 11  recovered              10000 non-null  int64  
dtypes: float64(2), int64(6), object(4)
memory usage: 937.6+ KB


In [8]:
df.isnull().sum()

,0
payment_id,0
customer_id,0
amount_inr,0
payment_method,0
failure_type,0
hour_of_day,0
minutes_since_failure,0
retry_count,0
previous_payments,0
previous_successes,0


In [7]:
df.describe()

,amount_inr,hour_of_day,minutes_since_failure,retry_count,previous_payments,previous_successes,previous_success_rate,recovered
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000
mean,2536.276373,11.539600,1084.176300,0.995700,7.051100,4.683200,0.668799,0.360100
std,2399.094353,6.915476,620.779125,0.818259,4.321985,3.453988,0.271634,0.480053
min,106.830000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1068.097500,6.000000,551.750000,0.000000,3.000000,2.000000,0.500000,0.000000
50%,1837.450000,11.000000,1087.000000,1.000000,7.000000,4.000000,0.714000,0.000000
75%,3136.805000,17.000000,1620.250000,2.000000,11.000000,7.000000,0.900000,1.000000
max,34253.690000,23.000000,2159.000000,2.000000,14.000000,14.000000,1.000000,1.000000


In [10]:
df['failure_type'].value_counts()

,count
failure_type,
checkout_abandoned,3457
card_declined,2545
insufficient_funds,1999
network_timeout,1999


In [11]:
df['recovered'].value_counts()

,count
recovered,
0,6399
1,3601


In [12]:
df.groupby('failure_type')['recovered'].mean()

,recovered
failure_type,
card_declined,0.258546
checkout_abandoned,0.391380
insufficient_funds,0.240120
network_timeout,0.555278


In [13]:
df.groupby("retry_count")["recovered"].mean()

,recovered
retry_count,
0,0.479668
1,0.362481
2,0.236621


In [16]:
df.groupby('previous_success_rate')['recovered'].mean().head(20)

,recovered
previous_success_rate,
0.000,0.157175
0.071,1.000000
0.077,1.000000
0.083,0.500000
0.091,0.200000
0.100,0.000000
0.111,0.083333
0.125,0.117647
0.143,0.303030


In [17]:
df.groupby('recovered')['amount_inr'].sum()

,amount_inr
recovered,
0,16264402.62
1,9098361.11


In [18]:
df.groupby("failure_type")["recovered"].agg(["count", "mean"])

,count,mean
failure_type,,
card_declined,2545,0.258546
checkout_abandoned,3457,0.391380
insufficient_funds,1999,0.240120
network_timeout,1999,0.555278


In [20]:
df["success_rate_group"] = pd.cut(
    df["previous_success_rate"],
    bins=[-0.01, 0.25, 0.50, 0.75, 1.0],
    labels=["0-25%", "25-50%", "50-75%", "75-100%"]
)

In [21]:
df.groupby("success_rate_group", observed=True)["recovered"].agg(["count", "mean"])

,count,mean
success_rate_group,,
0-25%,869,0.182969
25-50%,2165,0.276212
50-75%,3116,0.358151
75-100%,3850,0.448831


In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   payment_id             10000 non-null  object  
 1   customer_id            10000 non-null  object  
 2   amount_inr             10000 non-null  float64 
 3   payment_method         10000 non-null  object  
 4   failure_type           10000 non-null  object  
 5   hour_of_day            10000 non-null  int64   
 6   minutes_since_failure  10000 non-null  int64   
 7   retry_count            10000 non-null  int64   
 8   previous_payments      10000 non-null  int64   
 9   previous_successes     10000 non-null  int64   
 10  previous_success_rate  10000 non-null  float64 
 11  recovered              10000 non-null  int64   
 12  success_rate_group     10000 non-null  category
dtypes: category(1), float64(2), int64(6), object(4)
memory usage: 947.6+ KB


In [23]:
df["recovered"].value_counts(normalize=True)

,proportion
recovered,
0,0.6399
1,0.3601


In [24]:
features = [
    "amount_inr",
    "payment_method",
    "failure_type",
    "hour_of_day",
    "minutes_since_failure",
    "retry_count",
    "previous_payments",
    "previous_successes",
    "previous_success_rate"
]

In [25]:
target = "recovered"

In [26]:
X=df[features]
Y=df[target]

In [30]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    Y,
    test_size=0.2,
    random_state=42,
    stratify=Y
)

In [31]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

categorical_features = [
    "payment_method",
    "failure_type"
]

preprocessor = ColumnTransformer(
    transformers=[
        (
            "categorical",
            OneHotEncoder(handle_unknown="ignore"),
            categorical_features
        )
    ],
    remainder="passthrough"
)

In [34]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

model = Pipeline([
    ("preprocessor", preprocessor),
    ("classifier", LogisticRegression(max_iter=5000))
])

In [35]:
model.fit(X_train, y_train)

/usr/local/lib/python3.13/dist-packages/sklearn/compose/_column_transformer.py:1667: FutureWarning: 
The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).

  warnings.warn(


Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('categorical',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['payment_method',
                                                   'failure_type'])])),
                ('classifier', LogisticRegression(max_iter=5000))])

In [36]:
y_pred = model.predict(X_test)

In [37]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.86      0.79      1280
           1       0.64      0.44      0.52       720

    accuracy                           0.71      2000
   macro avg       0.69      0.65      0.66      2000
weighted avg       0.70      0.71      0.70      2000



In [38]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_pred))

[[1105  175]
 [ 404  316]]


In [39]:
from sklearn.metrics import accuracy_score

print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.7105


In [41]:
y_prob = model.predict_proba(X_test)

print(y_prob[:100])

[[0.90979151 0.09020849]
 [0.5541768  0.4458232 ]
 [0.21056513 0.78943487]
 [0.95345149 0.04654851]
 [0.95168315 0.04831685]
 [0.72002153 0.27997847]
 [0.84388739 0.15611261]
 [0.48607425 0.51392575]
 [0.8698042  0.1301958 ]
 [0.58762896 0.41237104]
 [0.74855847 0.25144153]
 [0.85796878 0.14203122]
 [0.58451748 0.41548252]
 [0.8837982  0.1162018 ]
 [0.56632317 0.43367683]
 [0.37321105 0.62678895]
 [0.82642187 0.17357813]
 [0.19190603 0.80809397]
 [0.90693416 0.09306584]
 [0.77547453 0.22452547]
 [0.62724684 0.37275316]
 [0.66481237 0.33518763]
 [0.87716461 0.12283539]
 [0.65324168 0.34675832]
 [0.4185566  0.5814434 ]
 [0.42133044 0.57866956]
 [0.64959755 0.35040245]
 [0.69645138 0.30354862]
 [0.66924656 0.33075344]
 [0.34934838 0.65065162]
 [0.79419482 0.20580518]
 [0.77888836 0.22111164]
 [0.33416521 0.66583479]
 [0.81695651 0.18304349]
 [0.57839318 0.42160682]
 [0.47072948 0.52927052]
 [0.79824208 0.20175792]
 [0.4264681  0.5735319 ]
 [0.75182692 0.24817308]
 [0.62121531 0.37878469]


In [43]:
recovery_probability = y_prob[:, 1]

print(recovery_probability[:10])

[0.09020849 0.4458232  0.78943487 0.04654851 0.04831685 0.27997847
 0.15611261 0.51392575 0.1301958  0.41237104]


In [44]:
results = X_test.copy()

results["actual_recovered"] = y_test.values
results["recovery_probability"] = recovery_probability
results["predicted_recovered"] = y_pred

results.head(10)

,amount_inr,payment_method,failure_type,hour_of_day,minutes_since_failure,retry_count,previous_payments,previous_successes,previous_success_rate,actual_recovered,recovery_probability,predicted_recovered
3791,837.56,netbanking,insufficient_funds,19,1535,1,3,1,0.333,0,0.090208,0
6511,753.78,card,card_declined,23,503,1,14,14,1.000,0,0.445823,0
1189,1485.18,card,checkout_abandoned,23,63,0,11,10,0.909,0,0.789435,1
5357,3078.38,card,insufficient_funds,2,2038,2,14,7,0.500,0,0.046549,0
5209,5683.55,netbanking,card_declined,2,1613,2,13,4,0.308,0,0.048317,0
6720,4658.51,card,insufficient_funds,19,1831,0,7,6,0.857,0,0.279978,0
9251,1431.02,card,insufficient_funds,7,1243,2,1,1,1.000,0,0.156113,0
436,2241.83,card,card_declined,20,919,0,7,7,1.000,1,0.513926,1
554,633.41,netbanking,card_declined,2,1794,1,9,6,0.667,0,0.130196,0
5152,656.22,netbanking,network_timeout,23,1940,1,14,9,0.643,0,0.412371,0


In [45]:
print("Baseline Model Performance")
print("--------------------------")
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2%}")

print("\nRecovery prediction performance:")
print(classification_report(y_test, y_pred))

Baseline Model Performance
--------------------------
Accuracy: 71.05%

Recovery prediction performance:
              precision    recall  f1-score   support

           0       0.73      0.86      0.79      1280
           1       0.64      0.44      0.52       720

    accuracy                           0.71      2000
   macro avg       0.69      0.65      0.66      2000
weighted avg       0.70      0.71      0.70      2000



In [46]:
threshold = 0.40

y_pred_40 = (recovery_probability >= threshold).astype(int)

print(classification_report(y_test, y_pred_40))

              precision    recall  f1-score   support

           0       0.78      0.74      0.76      1280
           1       0.57      0.62      0.60       720

    accuracy                           0.70      2000
   macro avg       0.67      0.68      0.68      2000
weighted avg       0.70      0.70      0.70      2000



# Day 2 — ML Baseline Complete

## Objective

Build a baseline machine learning model that estimates whether a
failed payment is likely to be recovered based on historical
payment behaviour.

## Dataset

- 10,000 historical payment records
- 12 features
- Failure types:
  - card_declined
  - insufficient_funds
  - network_timeout
  - checkout_abandoned

## Baseline Model

Logistic Regression with:
- One-hot encoding for categorical features
- Numerical features passed through directly
- 80/20 train-test split

## Results

Baseline threshold (50%):
- Accuracy: 71.05%
- Recovery precision: 64%
- Recovery recall: 44%
- Recovery F1-score: 52%

Experimental threshold (40%):
- Recovery precision: 57%
- Recovery recall: 62%
- Recovery F1-score: 60%

## Conclusion

The historical payment data contains useful predictive signals for
payment recovery.

The model outputs a recovery probability rather than directly
taking an action. This probability will be provided to the RevivePay
decision engine, which will combine it with business and safety
rules before deciding whether to retry, wait, escalate, or stop.

The Logistic Regression model is currently a baseline and can be
improved later if necessary.